nixos/modules/services/x11/xserver.nix

```diff
@@ -591,8 +591,23 @@ in
       let dmconf = cfg.displayManager;
           default = !(dmconf.gdm.enable
                     || dmconf.sddm.enable
-                    || dmconf.xpra.enable );
-      in mkIf (default) true;
+                    || dmconf.xpra.enable
+                    || dmconf.sx.enable
+                    || dmconf.startx.enable);
+      in mkIf (default) (mkDefault true);
+
+    # so that the service won't be enabled when only startx is used
+    systemd.services.display-manager.enable  =
+      let dmconf = cfg.displayManager;
+          nodmUsed = !(dmconf.gdm.enable
+                    || dmconf.sddm.enable
+                    || dmconf.xpra.enable
+                    || dmconf.lightdm.enable);
+      in mkIf (nodmUsed) (mkDefault false);
+
+    # error: The option `services.xserver.displayManager.job.execCmd' is used but not defined.
```

Comment:
You still know where this came from? Maybe we can fix it there quite easily without that hack.

Reply:
```
error: The option `services.xserver.displayManager.job.execCmd' is used but not defined.

       … while evaluating the attribute 'mergedValue'

       at /nix/store/w6fyzp0xqbh72q6w68hkfhfpbzzrzkw5-source/lib/modules.nix:612:5:

          611|     # Type-check the remaining definitions, and merge them. Or throw if no definitions.
          612|     mergedValue =
             |     ^
          613|       if isDefined then

       … while evaluating the option `services.xserver.displayManager.job.execCmd':

       … while evaluating the attribute 'value'

       at /nix/store/w6fyzp0xqbh72q6w68hkfhfpbzzrzkw5-source/lib/modules.nix:580:9:

          579|     in warnDeprecation opt //
          580|       { value = builtins.addErrorContext "while evaluating the option `${showOption loc}':" value;
             |         ^
          581|         inherit (res.defsFinal') highestPrio;

       … while evaluating anonymous lambda

       at /nix/store/w6fyzp0xqbh72q6w68hkfhfpbzzrzkw5-source/lib/modules.nix:166:72:

          165|           # For definitions that have an associated option
          166|           declaredConfig = mapAttrsRecursiveCond (v: ! isOption v) (_: v: v.value) options;
             |                                                                        ^
          167|

       … from call site

       at /nix/store/w6fyzp0xqbh72q6w68hkfhfpbzzrzkw5-source/lib/attrsets.nix:304:20:

          303|               then recurse (path ++ [name]) value
          304|               else f (path ++ [name]) value;
             |                    ^
          305|         in mapAttrs g set;

       … while evaluating 'g'

       at /nix/store/w6fyzp0xqbh72q6w68hkfhfpbzzrzkw5-source/lib/attrsets.nix:301:19:

          300|           g =
          301|             name: value:
             |                   ^
          302|             if isAttrs value && cond value

       … from call site

       … while evaluating the attribute 'value'

       at /nix/store/w6fyzp0xqbh72q6w68hkfhfpbzzrzkw5-source/lib/modules.nix:447:44:

          446|       defnsByName' = byName "config" (module: value:
          447|           [{ inherit (module) file; inherit value; }]
             |                                            ^
          448|         ) configs;

       … while evaluating 'dischargeProperties'

       at /nix/store/w6fyzp0xqbh72q6w68hkfhfpbzzrzkw5-source/lib/modules.nix:664:25:

          663|   */
          664|   dischargeProperties = def:
             |                         ^
          665|     if def._type or "" == "merge" then

       … from call site

       at /nix/store/w6fyzp0xqbh72q6w68hkfhfpbzzrzkw5-source/lib/modules.nix:593:137:

          592|         defs' = concatMap (m:
          593|           map (value: { inherit (m) file; inherit value; }) (builtins.addErrorContext "while evaluating definitions from `${m.file}':" (dischargeProperties m.value))
             |                                                                                                                                         ^
          594|         ) defs;

       … while evaluating definitions from `/nix/store/w6fyzp0xqbh72q6w68hkfhfpbzzrzkw5-source/nixos/modules/services/x11/xserver.nix':

       … while evaluating anonymous lambda

       at /nix/store/w6fyzp0xqbh72q6w68hkfhfpbzzrzkw5-source/lib/modules.nix:592:28:

          591|         # Process mkMerge and mkIf properties.
          592|         defs' = concatMap (m:
             |                            ^
          593|           map (value: { inherit (m) file; inherit value; }) (builtins.addErrorContext "while evaluating definitions from `${m.file}':" (dischargeProperties m.value))

       … from call site

       at /nix/store/w6fyzp0xqbh72q6w68hkfhfpbzzrzkw5-source/lib/modules.nix:592:17:

          591|         # Process mkMerge and mkIf properties.
          592|         defs' = concatMap (m:
             |                 ^
          593|           map (value: { inherit (m) file; inherit value; }) (builtins.addErrorContext "while evaluating definitions from `${m.file}':" (dischargeProperties m.value))

       … while evaluating the attribute 'values'

       at /nix/store/w6fyzp0xqbh72q6w68hkfhfpbzzrzkw5-source/lib/modules.nix:705:7:

          704|     in {
          705|       values = concatMap (def: if getPrio def == highestPrio then [(strip def)] else []) defs;
             |       ^
          706|       inherit highestPrio;

       … while evaluating the attribute 'values'

       at /nix/store/w6fyzp0xqbh72q6w68hkfhfpbzzrzkw5-source/lib/modules.nix:606:9:

          605|       in {
          606|         values = defs''';
             |         ^
          607|         inherit (defs'') highestPrio;

       … while evaluating the attribute 'mergedValue'

       at /nix/store/w6fyzp0xqbh72q6w68hkfhfpbzzrzkw5-source/lib/modules.nix:612:5:

          611|     # Type-check the remaining definitions, and merge them. Or throw if no definitions.
          612|     mergedValue =
             |     ^
          613|       if isDefined then

       … while evaluating the option `systemd.services.display-manager.script':

       … while evaluating the attribute 'value'

       at /nix/store/w6fyzp0xqbh72q6w68hkfhfpbzzrzkw5-source/lib/modules.nix:580:9:

          579|     in warnDeprecation opt //
          580|       { value = builtins.addErrorContext "while evaluating the option `${showOption loc}':" value;
             |         ^
          581|         inherit (res.defsFinal') highestPrio;

       … while evaluating anonymous lambda

       at /nix/store/w6fyzp0xqbh72q6w68hkfhfpbzzrzkw5-source/lib/modules.nix:166:72:

          165|           # For definitions that have an associated option
          166|           declaredConfig = mapAttrsRecursiveCond (v: ! isOption v) (_: v: v.value) options;
             |                                                                        ^
          167|

       … from call site

       at /nix/store/w6fyzp0xqbh72q6w68hkfhfpbzzrzkw5-source/lib/attrsets.nix:304:20:

          303|               then recurse (path ++ [name]) value
          304|               else f (path ++ [name]) value;
             |                    ^
          305|         in mapAttrs g set;

       … while evaluating 'g'

       at /nix/store/w6fyzp0xqbh72q6w68hkfhfpbzzrzkw5-source/lib/attrsets.nix:301:19:

          300|           g =
          301|             name: value:
             |                   ^
          302|             if isAttrs value && cond value

       … from call site

       … while evaluating the attribute 'condition'

       at /nix/store/w6fyzp0xqbh72q6w68hkfhfpbzzrzkw5-source/lib/modules.nix:748:14:

          747|     { _type = "if";
          748|       inherit condition content;
             |              ^
          749|     };

       … while evaluating the attribute 'condition'

       at /nix/store/w6fyzp0xqbh72q6w68hkfhfpbzzrzkw5-source/lib/modules.nix:748:14:

          747|     { _type = "if";
          748|       inherit condition content;
             |              ^
          749|     };

       … while evaluating 'dischargeProperties'

       at /nix/store/w6fyzp0xqbh72q6w68hkfhfpbzzrzkw5-source/lib/modules.nix:664:25:

          663|   */
          664|   dischargeProperties = def:
             |                         ^
          665|     if def._type or "" == "merge" then

       … from call site

       at /nix/store/w6fyzp0xqbh72q6w68hkfhfpbzzrzkw5-source/lib/modules.nix:593:137:

          592|         defs' = concatMap (m:
          593|           map (value: { inherit (m) file; inherit value; }) (builtins.addErrorContext "while evaluating definitions from `${m.file}':" (dischargeProperties m.value))
             |                                                                                                                                         ^
          594|         ) defs;

       … while evaluating definitions from `/nix/store/w6fyzp0xqbh72q6w68hkfhfpbzzrzkw5-source/nixos/modules/system/boot/systemd.nix':

       … while evaluating anonymous lambda

       at /nix/store/w6fyzp0xqbh72q6w68hkfhfpbzzrzkw5-source/lib/modules.nix:592:28:

          591|         # Process mkMerge and mkIf properties.
          592|         defs' = concatMap (m:
             |                            ^
          593|           map (value: { inherit (m) file; inherit value; }) (builtins.addErrorContext "while evaluating definitions from `${m.file}':" (dischargeProperties m.value))

       … from call site

       at /nix/store/w6fyzp0xqbh72q6w68hkfhfpbzzrzkw5-source/lib/modules.nix:592:17:

          591|         # Process mkMerge and mkIf properties.
          592|         defs' = concatMap (m:
             |                 ^
          593|           map (value: { inherit (m) file; inherit value; }) (builtins.addErrorContext "while evaluating definitions from `${m.file}':" (dischargeProperties m.value))

       … while evaluating the attribute 'values'

       at /nix/store/w6fyzp0xqbh72q6w68hkfhfpbzzrzkw5-source/lib/modules.nix:705:7:

          704|     in {
          705|       values = concatMap (def: if getPrio def == highestPrio then [(strip def)] else []) defs;
             |       ^
          706|       inherit highestPrio;

       … while evaluating the attribute 'values'

       at /nix/store/w6fyzp0xqbh72q6w68hkfhfpbzzrzkw5-source/lib/modules.nix:606:9:

          605|       in {
          606|         values = defs''';
             |         ^
          607|         inherit (defs'') highestPrio;

       … while evaluating the attribute 'mergedValue'

       at /nix/store/w6fyzp0xqbh72q6w68hkfhfpbzzrzkw5-source/lib/modules.nix:612:5:

          611|     # Type-check the remaining definitions, and merge them. Or throw if no definitions.
          612|     mergedValue =
             |     ^
          613|       if isDefined then

       … while evaluating the option `systemd.services.display-manager.serviceConfig':

       … while evaluating the attribute 'value'

       at /nix/store/w6fyzp0xqbh72q6w68hkfhfpbzzrzkw5-source/lib/modules.nix:580:9:

          579|     in warnDeprecation opt //
          580|       { value = builtins.addErrorContext "while evaluating the option `${showOption loc}':" value;
             |         ^
          581|         inherit (res.defsFinal') highestPrio;

       … while evaluating anonymous lambda

       at /nix/store/w6fyzp0xqbh72q6w68hkfhfpbzzrzkw5-source/lib/modules.nix:166:72:

          165|           # For definitions that have an associated option
          166|           declaredConfig = mapAttrsRecursiveCond (v: ! isOption v) (_: v: v.value) options;
             |                                                                        ^
          167|

       … from call site

       at /nix/store/w6fyzp0xqbh72q6w68hkfhfpbzzrzkw5-source/lib/attrsets.nix:304:20:

          303|               then recurse (path ++ [name]) value
          304|               else f (path ++ [name]) value;
             |                    ^
          305|         in mapAttrs g set;

       … while evaluating 'g'

       at /nix/store/w6fyzp0xqbh72q6w68hkfhfpbzzrzkw5-source/lib/attrsets.nix:301:19:

          300|           g =
          301|             name: value:
             |                   ^
          302|             if isAttrs value && cond value

       … from call site

       … while evaluating 'optional'

       at /nix/store/w6fyzp0xqbh72q6w68hkfhfpbzzrzkw5-source/lib/lists.nix:254:20:

          253|   */
          254|   optional = cond: elem: if cond then [elem] else [];
             |                    ^
          255|

       … from call site

       at /nix/store/w6fyzp0xqbh72q6w68hkfhfpbzzrzkw5-source/nixos/modules/system/boot/systemd.nix:917:16:

          916|             concatLists [
          917|               (optional (type == "oneshot" && (restart == "always" || restart == "on-success"))
             |                ^
          918|                 "Service '${name}.service' with 'Type=oneshot' cannot have 'Restart=always' or 'Restart=on-success'"

       … while evaluating anonymous lambda

       at /nix/store/w6fyzp0xqbh72q6w68hkfhfpbzzrzkw5-source/nixos/modules/system/boot/systemd.nix:910:16:

          909|       mapAttrsToList
          910|         (name: service:
             |                ^
          911|           let

       … from call site

       at /nix/store/w6fyzp0xqbh72q6w68hkfhfpbzzrzkw5-source/lib/attrsets.nix:259:16:

          258|   mapAttrsToList = f: attrs:
          259|     map (name: f name attrs.${name}) (attrNames attrs);
             |                ^
          260|

       … while evaluating anonymous lambda

       at /nix/store/w6fyzp0xqbh72q6w68hkfhfpbzzrzkw5-source/lib/attrsets.nix:259:10:

          258|   mapAttrsToList = f: attrs:
          259|     map (name: f name attrs.${name}) (attrNames attrs);
             |          ^
          260|

       … from call site

       … while evaluating the attribute 'value'

       at /nix/store/w6fyzp0xqbh72q6w68hkfhfpbzzrzkw5-source/lib/modules.nix:447:44:

          446|       defnsByName' = byName "config" (module: value:
          447|           [{ inherit (module) file; inherit value; }]
             |                                            ^
          448|         ) configs;

       … while evaluating 'dischargeProperties'

       at /nix/store/w6fyzp0xqbh72q6w68hkfhfpbzzrzkw5-source/lib/modules.nix:664:25:

          663|   */
          664|   dischargeProperties = def:
             |                         ^
          665|     if def._type or "" == "merge" then

       … from call site

       at /nix/store/w6fyzp0xqbh72q6w68hkfhfpbzzrzkw5-source/lib/modules.nix:593:137:

          592|         defs' = concatMap (m:
          593|           map (value: { inherit (m) file; inherit value; }) (builtins.addErrorContext "while evaluating definitions from `${m.file}':" (dischargeProperties m.value))
             |                                                                                                                                         ^
          594|         ) defs;

       … while evaluating definitions from `/nix/store/w6fyzp0xqbh72q6w68hkfhfpbzzrzkw5-source/nixos/modules/system/boot/systemd.nix':

       … while evaluating anonymous lambda

       at /nix/store/w6fyzp0xqbh72q6w68hkfhfpbzzrzkw5-source/lib/modules.nix:592:28:

          591|         # Process mkMerge and mkIf properties.
          592|         defs' = concatMap (m:
             |                            ^
          593|           map (value: { inherit (m) file; inherit value; }) (builtins.addErrorContext "while evaluating definitions from `${m.file}':" (dischargeProperties m.value))

       … from call site

       at /nix/store/w6fyzp0xqbh72q6w68hkfhfpbzzrzkw5-source/lib/modules.nix:592:17:

          591|         # Process mkMerge and mkIf properties.
          592|         defs' = concatMap (m:
             |                 ^
          593|           map (value: { inherit (m) file; inherit value; }) (builtins.addErrorContext "while evaluating definitions from `${m.file}':" (dischargeProperties m.value))

       … while evaluating the attribute 'values'

       at /nix/store/w6fyzp0xqbh72q6w68hkfhfpbzzrzkw5-source/lib/modules.nix:705:7:

          704|     in {
          705|       values = concatMap (def: if getPrio def == highestPrio then [(strip def)] else []) defs;
             |       ^
          706|       inherit highestPrio;

       … while evaluating the attribute 'values'

       at /nix/store/w6fyzp0xqbh72q6w68hkfhfpbzzrzkw5-source/lib/modules.nix:606:9:

          605|       in {
          606|         values = defs''';
             |         ^
          607|         inherit (defs'') highestPrio;

       … while evaluating the attribute 'mergedValue'

       at /nix/store/w6fyzp0xqbh72q6w68hkfhfpbzzrzkw5-source/lib/modules.nix:612:5:

          611|     # Type-check the remaining definitions, and merge them. Or throw if no definitions.
          612|     mergedValue =
             |     ^
          613|       if isDefined then

       … while evaluating the option `warnings':

       … while evaluating the attribute 'value'

       at /nix/store/w6fyzp0xqbh72q6w68hkfhfpbzzrzkw5-source/lib/modules.nix:580:9:

          579|     in warnDeprecation opt //
          580|       { value = builtins.addErrorContext "while evaluating the option `${showOption loc}':" value;
             |         ^
          581|         inherit (res.defsFinal') highestPrio;

       … while evaluating anonymous lambda

       at /nix/store/w6fyzp0xqbh72q6w68hkfhfpbzzrzkw5-source/lib/modules.nix:166:72:

          165|           # For definitions that have an associated option
          166|           declaredConfig = mapAttrsRecursiveCond (v: ! isOption v) (_: v: v.value) options;
             |                                                                        ^
          167|

       … from call site

       at /nix/store/w6fyzp0xqbh72q6w68hkfhfpbzzrzkw5-source/lib/attrsets.nix:304:20:

          303|               then recurse (path ++ [name]) value
          304|               else f (path ++ [name]) value;
             |                    ^
          305|         in mapAttrs g set;

       … while evaluating 'g'

       at /nix/store/w6fyzp0xqbh72q6w68hkfhfpbzzrzkw5-source/lib/attrsets.nix:301:19:

          300|           g =
          301|             name: value:
             |                   ^
          302|             if isAttrs value && cond value

       … from call site

       … while evaluating 'fold''

       at /nix/store/w6fyzp0xqbh72q6w68hkfhfpbzzrzkw5-source/lib/lists.nix:55:15:

           54|       len = length list;
           55|       fold' = n:
             |               ^
           56|         if n == len

       … from call site

       at /nix/store/w6fyzp0xqbh72q6w68hkfhfpbzzrzkw5-source/lib/lists.nix:59:8:

           58|         else op (elemAt list n) (fold' (n + 1));
           59|     in fold' 0;
             |        ^
           60|

       … while evaluating 'foldr'

       at /nix/store/w6fyzp0xqbh72q6w68hkfhfpbzzrzkw5-source/lib/lists.nix:52:20:

           51|   */
           52|   foldr = op: nul: list:
             |                    ^
           53|     let

       … from call site

       at /nix/store/w6fyzp0xqbh72q6w68hkfhfpbzzrzkw5-source/lib/trivial.nix:330:33:

          329|
          330|   showWarnings = warnings: res: lib.foldr (w: x: warn w x) res warnings;
             |                                 ^
          331|

       … while evaluating 'showWarnings'

       at /nix/store/w6fyzp0xqbh72q6w68hkfhfpbzzrzkw5-source/lib/trivial.nix:330:28:

          329|
          330|   showWarnings = warnings: res: lib.foldr (w: x: warn w x) res warnings;
             |                            ^
          331|

       … from call site

       at /nix/store/w6fyzp0xqbh72q6w68hkfhfpbzzrzkw5-source/nixos/modules/system/activation/top-level.nix:137:10:

          136|     then throw "\nFailed assertions:\n${concatStringsSep "\n" (map (x: "- ${x}") failedAssertions)}"
          137|     else showWarnings config.warnings baseSystem;
             |          ^
          138|

       … while evaluating 'fold''

       at /nix/store/w6fyzp0xqbh72q6w68hkfhfpbzzrzkw5-source/lib/lists.nix:55:15:

           54|       len = length list;
           55|       fold' = n:
             |               ^
           56|         if n == len

       … from call site

       at /nix/store/w6fyzp0xqbh72q6w68hkfhfpbzzrzkw5-source/lib/lists.nix:59:8:

           58|         else op (elemAt list n) (fold' (n + 1));
           59|     in fold' 0;
             |        ^
           60|

       … while evaluating 'foldr'

       at /nix/store/w6fyzp0xqbh72q6w68hkfhfpbzzrzkw5-source/lib/lists.nix:52:20:

           51|   */
           52|   foldr = op: nul: list:
             |                    ^
           53|     let

       … from call site

       at /nix/store/w6fyzp0xqbh72q6w68hkfhfpbzzrzkw5-source/nixos/modules/system/activation/top-level.nix:140:12:

          139|   # Replace runtime dependencies
          140|   system = foldr ({ oldDependency, newDependency }: drv:
             |            ^
          141|       pkgs.replaceDependency { inherit oldDependency newDependency drv; }

       … while evaluating the attribute 'system.build.toplevel'

       at /nix/store/w6fyzp0xqbh72q6w68hkfhfpbzzrzkw5-source/nixos/modules/system/activation/top-level.nix:309:5:

          308|
          309|     system.build.toplevel = system;
             |     ^
          310|

       … while evaluating the attribute 'buildCommand' of the derivation 'nixos-vm'

       at /nix/store/w6fyzp0xqbh72q6w68hkfhfpbzzrzkw5-source/pkgs/stdenv/generic/make-derivation.nix:205:7:

          204|     // (lib.optionalAttrs (attrs ? name || (attrs ? pname && attrs ? version)) {
          205|       name =
             |       ^
          206|         let
```

Reply:
seems that it is this https://github.com/NixOS/nixpkgs/blob/75b645529450fd4bac3b897dc1fab1701ade3465/nixos/modules/services/x11/xserver.nix#L705

Reply:
i've added a workaround and a fixme to that line for now but to fix it properly we need to move declaring the systemd service to a separate mkIf